lib/linked_list.rb

```diff
@@ -3,86 +3,200 @@
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      min = current.data
+      while current
+        if current.data < min
+          min = current.data
+        end
+        if current.next
+          current = current.next
+        else
+          return min
+        end
+      end
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      return 0 if @head.nil?
+      current = @head
+      count = 1
+      while current.next
+        count += 1
+        current = current.next
+      end
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      return nil if @head.nil?
+      return @head.data if index == 0
+      current = @head
+      i = 0
+      while i != index
+        if current.next
+          current = current.next
+          i += 1
+        else
+          return nil
+        end
+      end
+      return current.data
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def visit
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      while current
+        puts current.data
+        if current.next
+          current = current.next
+        else
+          return
+        end
+      end
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(value)
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      while current
+        if current.data == value
+          if current == @tail
+            goes_before = current.previous
+            goes_before.next = nil
+            @tail = goes_before
+            return
+          end
+          if current == @head
+            goes_after = current.next
+            current.previous = nil
+            @head = goes_after
+            return
+          end
+          goes_before = @previous
+          puts goes_before.data
+          goes_after = current.next
+          puts goes_after.data
+          goes_before.next = goes_after          

```

Comment:
Don't you also need to set 

```suggestion
          goes_before.next = goes_after          
          goes_after.previous = goes_before
```